src/nupic/bindings/sparse_matrix.i

```diff
@@ -940,14 +940,35 @@ def __div__(self, other):
     self->clipRowsBelowAndAbove(i.begin(), i.end(), a, b);
   }
 
-  // Returns the number of non-zeros per row, for all rows
-  PyObject* nNonZerosPerRow() const
+
+  %pythoncode %{
+    def nNonZerosPerRow(self, rows=None):
+      if rows is None:
+        return self._nNonZerosPerRow_allRows()
+      else:
+        return self._nNonZerosPerRow(numpy.asarray(rows, dtype="uint32"))
+  %}
+
+  PyObject* _nNonZerosPerRow_allRows() const
   {
     nupic::NumpyVectorT<nupic::UInt ## N1> nnzpr(self->nRows());
     self->nNonZerosPerRow(nnzpr.begin());
     return nnzpr.forPython();
   }
 
+  PyObject* _nNonZerosPerRow(PyObject* py_rows)
```

Comment:
The parts that are reading inputs don't do any allocation, so they can't leak. The returned vector is allocated by `NumpyVectorT`. In the `NumpyVectorT` destructor it calls Py_CLEAR on the Python object, releasing it if the refcount hits 0. In the `forPython` method it increments the refcount, so it relies on the python runtime to release this object later.

I just now experimented. I ran these methods millions of times and verified that the Python process's memory doesn't increase.